File: semana_4_5_patrones_creacionales/codigo_ejemplos/ejemplo3_builder.py
# Ejemplo 3: Builder en Python (EXACTO DEL PDF)

class Computadora:
    def __init__(self):
        self.cpu = None
        self.ram = None
        self.gpu = None

class Builder:
    def __init__(self):
        self.computadora = Computadora()
    
    def add_cpu(self, cpu):
        self.computadora.cpu = cpu
        return self
    
    def add_ram(self, ram):
        self.computadora.ram = ram
        return self
    
    def add_gpu(self, gpu):
        self.computadora.gpu = gpu
        return self
    
    def build(self):
        return self.computadora

pc = Builder().add_cpu("Intel i9").add_ram("32GB").add_gpu("RTX 4090").build()
print(vars(pc))

File: semana_4_5_patrones_creacionales/codigo_ejemplos/ejemplo3_builder.cpp
// Ejemplo 3: Builder en C++ (VERSIÓN CRUZADA del ejemplo Python)

#include <iostream>
#include <string>
using namespace std;

class Computadora {
public:
    string cpu;
    string ram;
    string gpu;
    
    void mostrar() {
        cout << "CPU: " << cpu << endl;
        cout << "RAM: " << ram << endl;
        cout << "GPU: " << gpu << endl;
    }
};

class Builder {
private:
    Computadora* computadora;
    
public:
    Builder() {
        computadora = new Computadora();
    }
    
    Builder* add_cpu(string cpu) {
        computadora->cpu = cpu;
        return this;
    }
    
    Builder* add_ram(string ram) {
        computadora->ram = ram;
        return this;
    }
    
    Builder* add_gpu(string gpu) {
        computadora->gpu = gpu;
        return this;
    }
    
    Computadora* build() {
        return computadora;
    }
};

int main() {
    Computadora* pc = (new Builder())
        ->add_cpu("Intel i9")
        ->add_ram("32GB")
        ->add_gpu("RTX 4090")
        ->build();
    
    pc->mostrar();
    return 0;
}
